### PR TITLE
Snapshot service to handle failed snapshots

### PR DESCRIPTION
**Pull request by patrickmann**
/nocl see associated enterprise PR

Part of Graylog2/graylog-plugin-enterprise#8233 

/jpd Graylog2/graylog-plugin-enterprise#8390

## Description
- Adds a button to delete a failed snapshot to the maintenance dropdown on the index set page.
- Clicking the button shows a confirm modal with further information on the impact of this process.

This feature relies on the attributes as part of the response of `GET /system/indices/index_sets/ID`:
**Example response:**
``` json
{
  id: "234asd2",
  title: "foo",
  index_prefix: "bar",
  has_failed_snapshot: true,
  failed_snapshot_name: "foor_warm"
}
```

- `has_failed_snapshot` (boolean) to determine whether to display the button
- `failed_snapshot_name` (optional string) to display the name of the index to delete in the confirm modal

It uses the following call to delete the snapshot:
`DELETE /api/system/indices/index_sets/INDEX_SET_ID/failed_snapshot`

## Motivation
In an Enterprise installation with warm tier, we may need to delete failed snapshots. 
This PR provides the necessary binding in Open.

## Screenshots
Screenshot 2024-08-29 at 12 11 56
Screenshot 2024-08-29 at 12 12 01

### CODE REVIEW

**Review by gally47 — APPROVED** [review]
Frontend LGTM.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/responses/IndexSetSummary.java [hunk 1]
```diff
@@ -67,17 +69,24 @@ public static IndexSetSummary create(@JsonProperty("id") @Nullable String id,
...
         return new AutoValue_IndexSetSummary(shards, replicas, rotationStrategyClass, rotationStrategy, retentionStrategyClass,
                 retentionStrategy, dataTiering, id, title, description, isDefault, canBeDefault, isWritable, indexPrefix,
                 creationDate, indexAnalyzer, indexOptimizationMaxNumSegments, indexOptimizationDisabled, fieldTypeRefreshInterval,
-                Optional.ofNullable(templateType), fieldTypeProfile, Objects.isNull(useLegacyRotation) || useLegacyRotation);
+                Optional.ofNullable(templateType), fieldTypeProfile, Objects.isNull(useLegacyRotation) || useLegacyRotation,
+                hasFailedSnapshot, failedSnapshotName);
     }
 
     public static IndexSetSummary fromIndexSetConfig(IndexSetConfig indexSet, boolean isDefault) {
+        return fromIndexSetConfig(indexSet, isDefault, Optional.empty());
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/responses/IndexSetSummary.java:86]: If I'm not mistaken, existing calls, e.g. the API list call calls, if they still use this method, will now potentially have different values for the snapshot fields of the returned `IndexSetSummary` than the calls for an individual summary when using the new method. I think this can be misleading.

E.g. `hasFailedSnapshot == false` could mean that there is no failed snapshot but also mean that we didn't even check. 

I think it's a not ideal that the snapshot information is placed on the top level of `IndexSetSummary`. Otherwise we could easily signal, e.g. in a list context, that the snapshot information is missing by omitting the value for the top level field.

What do you think about adding an optional top level field like `dataTieringStatus` or `dataTieringInfo` (preferably something more meaningful even), and putting the failure and/or snapshot information in there? Then the absence of the top level field means that the info is not available.

File: graylog2-server/src/test/java/org/graylog2/rest/resources/system/indexer/IndexSetsResourceTest.java [hunk 2]
```diff
@@ -155,33 +156,14 @@ public void list0() {
...
-        when(indexSetService.get("id")).thenReturn(Optional.of(indexSetConfig));
-
-        final IndexSetSummary summary = indexSetsResource.get("id");
-
-        verify(indexSetService, times(1)).get("id");
-        verify(indexSetService, times(1)).getDefault();
-        verifyNoMoreInteractions(indexSetService);
-        assertThat(summary).isEqualTo(IndexSetSummary.fromIndexSetConfig(indexSetConfig, false));
-    }
-
```
[the remark this answers is not in the document]
Comment (patrickmann) [reply graylog2-server/src/test/java/org/graylog2/rest/resources/system/indexer/IndexSetsResourceTest.java:170]: I considered expanding it for the new feature. But all it does is check that certain methods are called. There is no validation of any return values. Just doesn't seem to add much value for the effort required to maintain and update it.

**Review by kroepke — CHANGES_REQUESTED** [review]
Should the UI code not live next to the enterprise code and the main code allow for extension points?

As it stands we have code for a closed source feature in the open repository, which we avoid doing.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java [hunk 3]
```diff
@@ -206,9 +210,12 @@ public IndexSetStats globalStats() {
     public IndexSetSummary get(@ApiParam(name = "id", required = true)
                                @PathParam("id") String id) {
         checkPermission(RestPermissions.INDEXSETS_READ, id);
+        final IndexSet indexSet = indexSetRegistry.get(id).orElseThrow(() -> new NotFoundException("Couldn't find index set with ID <" + id + ">"));
         final IndexSetConfig defaultIndexSet = indexSetService.getDefault();
         return indexSetService.get(id)
-                .map(config -> IndexSetSummary.fromIndexSetConfig(config, config.equals(defaultIndexSet)))
+                .map(config -> IndexSetSummary.fromIndexSetConfig(
+                        config, config.equals(defaultIndexSet),
+                        snapshotService.getFailedSnapshotName(indexSet, config)))
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:218]: It seems to me this is the reason for all of the rest of the code to be in graylog2-server.

A better solution would be to create a more generic extension point for contributing index set information from plugins, instead of pulling the very specific snapshot implementation into the server code base.
As Othello noted, a more generic data tiering info endpoint that's empty by default is probably the better choice.

Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:218]: > a more generic data tiering info endpoint that's empty by default 


@kay Makes sense. Can we make this refactoring an issue for cooldown? 
We're running out of time to merge this in the current cycle and I hate rushing it.

Comment (kroepke) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:218]: I'd greatly prefer not to merge code we don't intend to keep, to be honest.

Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:218]: Cool - to expedite this, it'd be very helpful if you could spec this generic endpoint.

Comment (kroepke) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:218]: The mechanics are the same as the warm-tier configuration, basically a dummy implementation like SnapshotServiceNoop.
There's multiple options here:
* Either include a more generic property in IndexSetSummary, e.g. `data_tiering` and return a specific hot/warm object by calling data tiering summary implementation. The the hot/warm specific UI can use that to render the button if necessary and call the enterprise endpoint for deleting the snapshot, or whatever it needs to do. This way it's separated for any future data tiering implementation without any additional work required in the server
* Or entirely do it from the UI, loading (possibly in bulk) hot/warm details for each index set, that then returns the specific object, e.g. in this case the list of failed/partial snapshot information to let the UI render it.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/responses/DataTieringStatus.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.rest.resources.system.indexer.responses;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record DataTieringStatus(
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/responses/DataTieringStatus.java:21]: This should be empty and generally use `@JsonTypeInfo`, `@JsonSubTypes` annotations to add additional implementations to be returned.
This case should be a little easier, as we don't need to receive data from the API, we are simply returning information from the enterprise plugin, so you can potentially get away without using JsonSubType annotations and simply rely on Jackson to properly serialize it.
It's good practice to include a `type` property with a string value (e.g. `snapshot_status`) to allow for detecting which data tiering status object we have in the UI or other clients.

tl;dr: make this an interface, return concrete snapshot status class from the enterprise code's SnapshotDataTieringStatusService class.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java [hunk 5]
```diff
@@ -362,4 +369,23 @@ public void delete(@ApiParam(name = "id", required = true)
...
+    })
+    public void delete(@ApiParam(name = "id", required = true)
+                       @PathParam("id") String id) {
+        checkPermission(RestPermissions.INDEXSETS_EDIT, id);
+
+        tieringStatusService.deleteFailedSnapshot(
+                indexSetRegistry.get(id).orElseThrow(() -> new NotFoundException("Failed to get index set <" + id + ">")),
+                indexSetService.get(id).orElseThrow(() -> new NotFoundException("Failed to get config for index set <" + id + ">"))
+        );
+    }
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexSetsResource.java:390]: this needs to live in the enterprise code base

**Review by kroepke — APPROVED** [review]
thx for addressing this!

File: graylog2-web-interface/src/components/indices/IndicesMaintenanceDropdown.tsx [hunk 6]
```diff
@@ -54,6 +58,9 @@ const IndicesMaintenanceDropdown = ({ indexSet, indexSetId }: Props) => {
       <DropdownButton bsStyle="info" title="Maintenance" id="indices-maintenance-actions" pullRight>
         <MenuItem eventKey="1" onClick={onRecalculateIndexRange}>Recalculate index ranges</MenuItem>
         {cycleButton}
+        {indexSet?.data_tiering_status?.has_failed_snapshot && dataTieringPlugin && (
+          <dataTieringPlugin.DeleteFailedSnapshotMenuItem eventKey="3" indexSetId={indexSetId} />
+        )}
```
Comment (kroepke) [inline graylog2-web-interface/src/components/indices/IndicesMaintenanceDropdown.tsx:63]: Ideally, this would also not live in the server, but for the sake of moving this forward let's leave it like this.

File: graylog2-web-interface/src/components/indices/data-tiering/types.ts [hunk 7]
```diff
@@ -43,6 +43,11 @@ export type DataTieringFormValues = {
   warm_tier_repository_name?: string | null,
 }
 
+export type DataTieringStatus = {
+  has_failed_snapshot: boolean,
+  failed_snapshot_name: string | null,
+}
+
```
Comment (kroepke) [inline graylog2-web-interface/src/components/indices/data-tiering/types.ts:50]: Ideally, this would also not live in the server, but for the sake of moving this forward let's leave it like this.
